### PR TITLE
Patch : Refactors the TrySerializeValueParameter method for PatchOperation and makes it public

### PR DESCRIPTION
**Pull request by anujtoshniwal**
# Pull Request Template

## Description

Exposes the TrySerializeValueParameter method for PatchOperation so that it can be consumed in Encryption package. Need access to the serialized value parameter so that it can be encrypted according to the client encryption policy defined by the user.

## Type of change

- [] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Review by kr-santosh — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Patch/PatchOperationsJsonConverter.cs [hunk 1]
```diff
@@ -87,8 +88,17 @@ public override bool CanConvert(Type objectType)
...
+                    using (valueStream)
+                    {
+                        using (StreamReader streamReader = new StreamReader(valueStream))
+                        {
+                            valueParam = streamReader.ReadToEnd();
+                        }
+                    }
+
                     writer.WritePropertyName(PatchConstants.PropertyNames.Value);
                     writer.WriteRawValue(valueParam);
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Patch/PatchOperationsJsonConverter.cs:103]: Isn't position set to the start a contract requirement?

**Review by kirankumarkolli — CHANGES_REQUESTED** [review]
Waiting for clarify on stream contract

File: Microsoft.Azure.Cosmos/src/Patch/PatchOperationsJsonConverter.cs (hunk 1 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos/src/Patch/PatchOperationsJsonConverter.cs:103]: Yes, that is expected.

File: Microsoft.Azure.Cosmos/src/Patch/PatchOperation.cs [hunk 2]
```diff
@@ -29,9 +30,16 @@ abstract class PatchOperation
...
 
-        internal virtual bool TrySerializeValueParameter(
+        /// <summary>
+        /// Serializes the value parameter, if specified for the PatchOperation.
+        /// </summary>
+        /// <param name="cosmosSerializer">Serializer to be used.</param>
+        /// <param name="valueParam">Outputs the serialized stream if value parameter is specified, null otherwise.</param>
+        /// <returns>True if value is serialized, false otherwise.</returns>
+        /// <remarks>Output stream should be disposed after use.</remarks>
+        public virtual bool TrySerializeValueParameter(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Patch/PatchOperation.cs:40]:
```
#if PREVIEW
    public
#else 
    internal
#endif
```

File: Microsoft.Azure.Cosmos/src/Patch/PatchOperationCore{T}.cs [hunk 3]
```diff
@@ -33,20 +33,22 @@ internal sealed class PatchOperationCore<T> : PatchOperation<T>
 
         public override string Path { get; }
 
-        internal override bool TrySerializeValueParameter(
+        public override bool TrySerializeValueParameter(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Patch/PatchOperationCore{T}.cs:36]: Please add preview flag

File: Microsoft.Azure.Cosmos/src/Patch/PatchOperationsJsonConverter.cs [hunk 4]
```diff
@@ -87,8 +88,17 @@ public override bool CanConvert(Type objectType)
...
 
-                if (operation.TrySerializeValueParameter(this.userSerializer, out string valueParam))
+                if (operation.TrySerializeValueParameter(this.userSerializer, out Stream valueStream))
                 {
+                    string valueParam;
+                    using (valueStream)
+                    {
+                        using (StreamReader streamReader = new StreamReader(valueStream))
+                        {
+                            valueParam = streamReader.ReadToEnd();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Patch/PatchOperationsJsonConverter.cs:98]: This is a blocking call. Any way to use async?

Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos/src/Patch/PatchOperationsJsonConverter.cs:98]: Don't think so. We need the string at this point to construct the request body. 
This isn't new code, have just moved it from TrySerializeValueParameter method implementation.

**Review by j82w — APPROVED** [review]

**Review by kirankumarkolli — APPROVED** [review]

**Pull request merged**